Avoid fork on mount for overlay2 in common case

dmcgowan opened the pull request:
In the common case where the user is using /var/lib/docker and
an image with less than 60 layers, forking is not needed. Calculate
whether absolute paths can be used and avoid forking to mount in
those cases.

I do not have any statistics of what percentage of images
are less than 60 layers deep, but my guess is a vast majority and
even larger majority when accounting for images which are run most often.
## Mount Benchmarks
### Overlay2 (pre-change) vs Overlay2 (updated)

```
benchmark                         old ns/op     new ns/op     delta
BenchmarkGetSingleBaseMount-8     7744409       231444        -97.01%
BenchmarkGet20BaseMount-8         8324308       307961        -96.30%
BenchmarkGet50BaseMount-8         8482378       420499        -95.04%
BenchmarkGet100BaseMount-8        8252170       7976995       -3.33%
```
### Overlay vs Overlay2 (updated)

```
benchmark                         old ns/op     new ns/op     delta
BenchmarkGetSingleBaseMount-8     339539        231444        -31.84%
BenchmarkGet20BaseMount-8         249545        307961        +23.41%
BenchmarkGet50BaseMount-8         200676        420499        +109.54%
BenchmarkGet100BaseMount-8        252296        7976995       +3061.76%
```
### AUFS vs Overlay2 (updated)

```
benchmark                         old ns/op     new ns/op     delta
BenchmarkGetSingleBaseMount-8     305157        231444        -24.16%
BenchmarkGet20BaseMount-8         1830297       307961        -83.17%
BenchmarkGet50BaseMount-8         10263958      420499        -95.90%
BenchmarkGet100BaseMount-8        36276899      7976995       -78.01%
```
### AUFS vs Overlay2 (pre-change)

```
benchmark                         old ns/op     new ns/op     delta
BenchmarkGetSingleBaseMount-8     305157        7744409       +2437.84%
BenchmarkGet20BaseMount-8         1830297       8324308       +354.81%
BenchmarkGet50BaseMount-8         10263958      8482378       -17.36%
BenchmarkGet100BaseMount-8        36276899      8252170       -77.25%
```

Note `overlay` will always mount faster than `overlay2` for a larger number of lower directories since it can only have one lower directory (which it achieves at the expense of duplicating files). These are the first benchmarks I have done with aufs vs overlay2 so I included the benchmark comparison with the pre-changed overlay2.

Benchmarks done with https://github.com/dmcgowan/dsdbench

tonistiigi commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -409,13 +409,43 @@ func (d *Driver) Get(id string, mountLabel string) (s string, err error) {
+	// name and path separator must be accounted for. The absolute
+	// directories for work and diff are also accounted for along with a
+	// buffer for other mount labels.
+	if (len(d.home)+idLength+len(linkDir)+2)*len(splitLowers)+len(dir)*2+512 < pageSize {
```
Would it make sense to make this calculation more precise? I think even calculating the whole `opts` even if it is not used would not be noticeably slow(good thing you have benchmarks to prove me wrong). If not then please add a comment that `512` is a buffer for the contants and label, I was initially confused by that.

dmcgowan replied:
I had considered just calculating the opts using absolute paths optimistically, I think this might make even more sense considering I found the threshold to be over 60 layers.

dmcgowan commented:
Updated to optimistically create the mount data with absolute paths and only use relative paths when the mount data is larger than a page. Tested to make sure that the extra buffer is not needed when the mount data is fully computed (see https://github.com/dmcgowan/dsdbench/commit/071a52647be33aa3feb959d6733b1560182e9f47#diff-c16a9df1a4a40f7564a79598949a5199R136).

AkihiroSuda commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -409,13 +409,36 @@ func (d *Driver) Get(id string, mountLabel string) (s string, err error) {
+		mountTarget = path.Join(id, "merged")
+	}
+
+	if len(mountData) > pageSize {
```
I think it will be much more readable if this branch is moved to L430. i.e.,
```go
mountData = label.FormatMountLabel(opts, mountLabel)
if len(mountData) > pageSize {
...
}
```

dmcgowan replied:
Agree, updated

dmcgowan committed:
Avoid fork on mount for overlay2 in common case

In the common case where the user is using /var/lib/docker and
an image with less than 60 layers, forking is not needed. Calculate
whether absolute paths can be used and avoid forking to mount in
those cases.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

tonistiigi commented:
LGTM

LK4D4 commented:
LGTM
